src/bullet_trace/BulletTrace.gd

```diff
@@ -0,0 +1,54 @@
+@icon("res://src/assets/noexport/artillery-shell.svg")
+class_name BulletTrace
+extends Line2D
+## Class for handling bullet trace animations, using the [Line2D] logic. [br]
+## An object of this class must always have a shader with a [code]lifetime[/code] and a [code]life[/code] variable.
+
+## Value dictating how far the bullet travels by default.
+## TODO: Make this variable weapon specific.
+static var ray_distance: float = 1000.0
+## Value dictating the scale of the random angle offset that occurs when a bullet is fired.
+## TODO: Make this variable weapon specific.
+static var ray_angle_offset: float = 0.04
+## Value dictating how long a bullet trace lingers before vanishing.
+## Should be scaled with the [code]ray_distance[/code] variable.
+static var trace_lifetime_factor: float = ray_distance * 2.0
+
+## Value dictating how long the bullet trace lingers.
+var lifetime: float
+## Value dictating how much time the bullet has left.
+var life: float
+
+func _process(delta: float) -> void:
+	life -= delta
+	material.set_shader_parameter("lifetime", lifetime)
+	material.set_shader_parameter("life", life)
+	if life <= 0.0:
+		queue_free()
+
+## Static function for spawning in bullet traces and dealing damage to the target. [br]
+## Takes [param source] as the source of the bullet, which is used to create the appropriate [DamageData]. [br]
+## Takes [param facing_angle] to orient the bullet in the correct direction. [br]
+## Takes [param damage] that gets passed into the [DamageData], if a target is hit.
+static func shoot_at(source: Node2D, facing_angle: float, damage: float) -> void:
```

Comment:
bullet traces should NOT be expected to pass damage etc. They should be purely visual effects which other things can use. Further, the only 2 pieces of data you NEED for a tracer is the start and end point, any other metadata *should* be to do with how they are rendered - line thickness, lifetime etc.